soc: cc13x2_cc26x2: P chips + Sub-GHz radio + CC1352P Launchpad support

niflostancu opened the pull request:
I added two new SOC definitions (for the P chip variants) and modified the IEEE802154 Sub-GHz driver to support high-power 20dBm TX (for CC1352P only).

I also created a new board (cc1352p_launchxl) with RF antenna switching implementation (based on RF event callbacks, as done in TI samples).
Tested using two of these boards and the echo client / server samples (as expected, RSSI grealty improved).

I will come later this week with an updated for 2.4GHz driver.

Request for comments (should I make a new issue?). Thanks.

niflostancu committed:
soc: cc13xx_cc26xx: config option to enable boost mode

Sets CCFG_FORCE_VDDR_HH to 1 to be able to use 14dBm TX power.

Signed-off-by: Stancu Florin <niflostancu@gmail.com>

niflostancu committed:
boards: cc1352p1_launchxl: new board, support antenna switching

Implements custom board antenna switching between 2.4GHz,
High-Power and Sub1GHz.

Signed-off-by: Stancu Florin <niflostancu@gmail.com>

niflostancu committed:
soc: cc13xx_cc26xx: add P chip variants to Kconfig

Add SOC_CC1352P and SOC_CC2652P chip types to SoC's Kconfig
(with integrated high power amplifiers).

Also requires modifications to HAL TI's family conditions.

Signed-off-by: Stancu Florin <niflostancu@gmail.com>

niflostancu committed:
soc: cc13xx_cc26xx: add Kconfig option for custom radio hwattrs

Useful for implementing board-specific antenna switching using the RF
event callback.

Note: requires updated hal_ti module.

Signed-off-by: Stancu Florin <niflostancu@gmail.com>

niflostancu committed:
drivers: ieee802154: cc13xx_cc26xx_subg: PA TX amplifier support

Enhance IEEE802154 Sub-GHz driver to support CC13x2P's internal power
amplifier (20dBm) for TX.

Signed-off-by: Stancu Florin <niflostancu@gmail.com>

cfriedt commented on boards/arm/cc1352p1_launchxl/Kconfig.board:
```diff
@@ -0,0 +1,21 @@
+# TI CC1352R LaunchXL board
+
+# Copyright (c) 2019 Brett Witherspoon
```
Please use your own identifier here

cfriedt commented on boards/arm/cc1352p1_launchxl/Kconfig.defconfig:
```diff
@@ -0,0 +1,11 @@
+# TI CC1352P1 LaunchXL board
+
+# Copyright (c) 2019 Brett Witherspoon
```
Same, please use your own identifier

cfriedt commented on boards/arm/cc1352p1_launchxl/board.cmake:
```diff
@@ -0,0 +1,5 @@
+# Copyright (c) 2019 Brett Witherspoon
```
Please use your own identifier

cfriedt commented on boards/arm/cc1352p1_launchxl/board_antenna.c:
```diff
@@ -0,0 +1,156 @@
+/*
+ * Copyright (c) 2018-2019, Texas Instruments Incorporated
```
Please use your own identifier here (unless you are an employee of TI - just going by your profile).

cfriedt commented on boards/arm/cc1352p1_launchxl/cc1352p1_launchxl.dts:
```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright (c) 2019 Brett Witherspoon
```
Please use your own identifier

cfriedt commented on boards/arm/cc1352p1_launchxl/cc1352p1_launchxl_defconfig:
```diff
@@ -0,0 +1,23 @@
+#
+# Copyright (c) 2019 Brett Witherspoon
```
Please use your own identifier

cfriedt commented on west.yml:
```diff
@@ -124,7 +124,7 @@ manifest:
         - hal
     - name: hal_ti
-      revision: 1992a4c536554c4f409c36896eda6abdc414d277
+      revision: pull/28/head
```
We will need to hold off on merging this until zephyrproject-rtos/hal_ti#28 is merged, and then this line will need to be updated with the appropriate revision.

cfriedt replied:
I would guess that this also will need zephyrproject-rtos/hal_ti#27

niflostancu replied:
Yes, you are right, I think it requires the updated HAL TI because of a `CMD_PROP_RADIO_SETUP_PA` addition (although it's an alias to the old radio setup command, I haven't tested the PA without the SDK update):
```c
#define CMD_PROP_RADIO_SETUP_PA                                 CMD_PROP_RADIO_SETUP
```

cfriedt commented:
I have one of these boards on my shelf atm, so I would be interested in giving this a shot. Thanks for adding support for it to Zephyr!

niflostancu commented:
Thanks for the review, will get to them next week (plus some enhancements).

You can test them by using a CC1352P for sending (e.g., echo client) and a CC1352R for receiving (you should see increased RSSI). Note that you need to slightly alter the echo example's `overlay-802154-subg.conf`, e.g.:
```conf
CONFIG_IEEE802154_CC13XX_CC26XX=n
CONFIG_IEEE802154_CC13XX_CC26XX_SUB_GHZ=y
CONFIG_NET_CONFIG_IEEE802154_DEV_NAME="IEEE802154_1"
CONFIG_NET_CONFIG_IEEE802154_CHANNEL=0  # 868MHz in the EU
CONFIG_NET_CONFIG_IEEE802154_RADIO_TX_POWER=14
```
(the default subg overlay is broken with multiple IEEE802154 interfaces, should I include the fix in the pull req?)

Note that the 2.4GHz driver doesn't have any of the PA changes, yet (planned in the future, maybe a different pull request).
I also need to check / update the documentation for the CC1352P variant of the board.

Pillar1989 commented:
Hi @naiyuantian    
I have done work on this before. I believe the code is 90% similar to yours. 
https://github.com/jadonk/zephyr/commit/98a4b8edd6477443cad0a11924e08b91367b33a3 

1， merge ieee802154_cc13xx_cc26xx_subg and ieee802154_cc13xx_cc26xx,  I think they should have been able to be merged and I'm curious as to why this wasn't done at the time? 
2， Find a suitable way to add more radio configuration settings.  Currently only one radio configuration is supported, which is clearly not good enough.

cc @jadonk

Pillar1989 commented:
@cfriedt  I've emailed you my thoughts before, it must be a problem with your email client, it's not presenting my words to you.

cfriedt commented:
> 1， merge ieee802154_cc13xx_cc26xx_subg and ieee802154_cc13xx_cc26xx, I think they should have been able to be merged and I'm curious as to why this wasn't done at the time?

@Pillar1989 - it's not super trivial to have multiple only slightly similar drivers in the same file. They do differ significantly. It would be best to factor-out any common code into a separate .c file that can be linked along with either driver.

> 2， Find a suitable way to add more radio configuration settings. Currently only one radio configuration is supported, which is clearly not good enough.

We have an issue for that  😃 - #29750

I would highly suggest starting out with an RFC / issue for multiple interface network config. I think DT would be a logical place for that, but it clearly is a lot of work and will need buy in from a number of people.

cc @rlubos, @jukkar

Pillar1989 commented:
@cfriedt  I think now is a very good time to do this. I'm currently working on the project BeagleConnect and I believe @niflostancu  is also working on CC13xx related projects. Maybe one of the maintainers of zephyr will build a framework first and we'll come to join more radio configuration settings.

niflostancu committed:
drivers: ieee802154: cc13xx_cc26xx_subg: custom RF setup kconfig

Possibility to use custom radio setup structure and/or custom
TX power tables via external module.

Useful if a board wants to change the RF modulation or protocol.

Signed-off-by: Stancu Florin <niflostancu@gmail.com>

niflostancu committed:
samples: net: echo: client/server: workaround for 802154-subg

Workaround for Zephyr not being able to configure multiple both
the 2.4GHz and sub-1ghz interfaces.

Signed-off-by: Stancu Florin <niflostancu@gmail.com>

niflostancu commented:
I updated my PR:
- Fixed the copyright comments;
- Added a new feature: Kconfig options to make the user / board able to override RADIO_SETUP command / TX power table entries using external module with TI RF Studio-generated values;
- Workaround to define CMD_PROP_RADIO_DIV_SETUP_PA (for the current HAL SDK version);
- Workaround for echo client/server samples' 802154 subg overlay.

Regarding the multiple interface configuration problem, I think it's best to handle it in a separate issue / pull request (hopefully, the workaround will do for now). OFC, I will help testing with my boards ;)

cfriedt commented on boards/arm/cc1352p1_launchxl/boosterpack_connector.dtsi:
```diff
@@ -0,0 +1,44 @@
+/*
```
if this is unchanged from `cc1352r_launchxl`, then just use a relative include directory. Same thing was done for the `cc1342r_sensortag`.
```
#include "../cc1352r1_launchxl/boosterpack_connector.dtsi"
```

cfriedt commented:
> > @Pillar1989 - it's not super trivial to have multiple only slightly similar drivers in the same file. They do differ significantly. It would be best to factor-out any common code into a separate .c file that can be linked along with either driver.
> 
> I believe the biggest role in doing this is using both 2.4G and Sub-1 GHz radios inside the same firmware. This is very important in a production environment.

The other main reason why 2.4 GHz and Sub-1 GHz should not be in the same file is that they can both be be active at the same time in a multi-phy scenario. In that case, it would be necessary to not use conditional compilation, and both instances would need to be built. If driver initialization were initiated from device tree, then it would make sense for each radio to have their own init macros at the bottom of separate files.

jadonk commented:
@cfriedt I'd like to find the shortest path to getting support for BeagleConnect Freedom in the mainline and a path to improve the SDK as quick and easy as possible. What do you see as the logical next step, in the simplest terms?

cfriedt commented:
@jadonk Definitely add the launchpad first, get all of the kinks worked out. Then the 'p' SoC will be integrated.

After that, it's mainly documentation, kconfig, device tree, etc. 

W.r.t. multiple phys / network interfaces, until there is a generic solution that everyone has bought into, it might be best to configure things at the board level for the BeagleConnect Freedom.

If you have some people who can do it work on #29750 it would be really helpful to put together an RFC first and tag the issues for dev review, and join in on the dev review call.

Likely won't get in for v2.7.0 because we're just merging bugfixes at this point, but should be able to get in for v3.0.0.

niflostancu commented:
Okay, so I tested the 2.4 radio using CC1352P launchpads and it works without any driver adaptations.
I also popped up Smart RF Studio and compared the overrides / div setup, no differences between CC1352R and CC1352P (because the power amplifier is only used for Sub-GHz, 2.4GHz still has the 5 dbm max tx limit probably because of regulations).
Note that sub-ghz was tested months ago (I have it in production and its working with no problems).  

I also made the requested change to include the `boosterpack_connector.dtsi` from the R1 board and removed the duplicate.
Is there anything else i should include into the PR?
I think it would be a good idea to port the suspend patch (#36697) to the 2.4 driver, too. Should I do that in here or as separate PR?
Thanks.

Echo test log (2.4GHz):
```txt
uart:~$ *** Booting Zephyr OS build v2.7.0-rc2-48-g6db180b98b7c  ***
[00:00:00.024,658] <inf> net_config: Initializing network
[00:00:00.132,995] <inf> net_config: IPv6 address: 2001:db8::2
[00:00:00.133,178] <inf> net_echo_client_sample: Run echo client
[00:00:00.133,209] <inf> net_echo_client_sample: Network connected
[00:00:00.133,239] <inf> net_echo_client_sample: Starting...
[00:00:00.138,305] <dbg> net_echo_client_sample.send_udp_data: IPv6 UDP: Sent 32 bytes
[00:00:00.188,507] <dbg> net_echo_client_sample.process_udp_proto: IPv6 UDP: Received and compared 32 bytes, all ok
[00:00:00.199,066] <dbg> net_echo_client_sample.send_udp_data: IPv6 UDP: Sent 89 bytes
[00:00:00.225,311] <dbg> net_echo_client_sample.process_udp_proto: IPv6 UDP: Received and compared 89 bytes, all ok
[00:00:00.229,248] <dbg> net_echo_client_sample.send_udp_data: IPv6 UDP: Sent 15 bytes
[00:00:00.235,565] <dbg> net_echo_client_sample.process_udp_proto: IPv6 UDP: Received and compared 15 bytes, all ok
[00:00:00.243,408] <dbg> net_echo_client_sample.send_udp_data: IPv6 UDP: Sent 93 bytes
```

cfriedt commented:
@niflostancu - can you please create a new PR? I'm able to merge things in hal/ti now there was no maintainer for some time, so PRs will not suffer from bitrot
